### DISCUSSION TITLE
How the Keyboard and mouse is mapped in vsg

### DISCUSSION THREAD

**Discussion by VasanthBalguri** [vsg-dev/VulkanSceneGraph#1090] [Q&A]
Hi Everyone,

I am currently working on wayland window, but I am facing some challenges while mapping buttons and keys to vsg events:
1. I can see that in vsg we use vsg::KeySymbol for keyboard keys, but for mouse buttons we directly send the value that we get from x11 xcb_button_press_event_t->detail. So 1 for Left mouse, 2 for Middle mouse, 3 for right mouse, but what about 5 button mouse, how to handle all buttons?
2. I cannot understand this part of code:
`vsg::clock::time_point event_time = _first_xcb_time_point + std::chrono::milliseconds(button_release->time - _first_xcb_timestamp);`

**Comment by VasanthBalguri** [comment]
Also I need some help regarding vsg::ButtonMask. There is no such concept in wayland, unlike X11, in what scenarios can we apply button mask?

**Reply by robertosfield** [reply]
Button masks has been a standard part of UI's for as long as I can remember.  It's essentially a state variable rather than an event one, so it tells you what all buttons are pressed at any time.

You might be able to emulate it, though ideally you'd want to Windowing system to tell you in case the user moves the mouse out of a window you lose event focus then the mouse comes back into focus but with different buttons pressed.

**Reply by VasanthBalguri** [reply]
Oh okay, this seems very different from how wayland works, but I want to know what can I send as button mask in mouse events?

**Comment by robertosfield** [comment]
> Hi Everyone,
> 
> I am currently working on wayland window, but I am facing some challenges while mapping buttons and keys to vsg events:
> 
>     1. I can see that in vsg we use vsg::KeySymbol for keyboard keys, but for mouse buttons we directly send the value that we get from x11 xcb_button_press_event_t->detail. So 1 for Left mouse, 2 for Middle mouse, 3 for right mouse, but what about 5 button mouse, how to handle all buttons?

I just tested vsginput on my Kubuntu 22.04/Xcb system and my 5 button mouse maps to button number 1,2, 3, 4 and 5.

So the Xcb_Window.cpp works OK so hopefully can help.

>     2. I cannot understand this part of code:
>        `vsg::clock::time_point event_time = _first_xcb_time_point + std::chrono::milliseconds(button_release->time - _first_xcb_timestamp);`

This is for adapting Xcb event times into a time frame relative to the rest of the VSG.  Perhsaps the following section might be of help:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/platform/xcb/Xcb_Window.cpp#L464

**Reply by VasanthBalguri** [reply]
Hi Robert,

For event_time, what should I choose as first_xcb_time_point and first_xcb_timestamp, is it when the window is open or when the vulkan surface is created?

**Reply by VasanthBalguri** [reply]
Also I want to know how the mouse buttons are mapped at vsg end, like what integers do I need to pass to vsg::mousepressevent button argument

**Reply by robertosfield** [reply]
It's 5 years since I wrote this code so I'm a bit cold on it.  I'd have to dig through all the code to figure it out.  

It might not even be the best way for your to do it for Wayland - the other platform/* combinations don't this trick, instead just use clock::now() when the event was handled by the VSG implement rather than try to compute the actual time of the event in the way I did for Xcb.  

FYI, I only wrote the Xcb_Window.cpp, the other platforms combinations were written by others.

**Comment by robertosfield** [comment]
> Also I want to know how the mouse buttons are mapped at vsg end, like what integers do I need to pass to vsg::mousepressevent button argument

Button 1 to button 1... etc. I don't know what more I can say.

Could you try out the vsginput example - it reports the event information so you should be able to see how how all the buttons are mapped and how the button mask changes.

**Comment by VasanthBalguri** [comment]
Thanks for your inputs @robertosfield , it definitely helped me, I will close this discussion for now.
